### DISCUSSION TITLE
why can not i use this rule to validate variable declartion

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1486] [Q&A] [answered]
why can not i use this rule to validate variable declaration
full langium file can be seen below demo.txt
[demo.txt](https://github.com/eclipse-langium/langium/files/15218959/demo.txt)

![image](https://github.com/eclipse-langium/langium/assets/26995040/3ce50001-5423-4d9f-9255-ac386ca18f61)

**Comment by msujew** [comment, marked as answer]
Hey @zrwsmd,

I assume you're using the default lexical scoping mechanism. Since the variables inside of the InputsListSingle object are nested in its local scope, they are not visible to the elements of the surrounding `PLC_PRG` program. You need to adjust the `ScopeComputation` (see [here for an example](https://langium.org/guides/scoping/qualified-name/)) to ensure that elements are visible outside of their parent node.

Note that using the word *validate* here is incredibly confusing. The issue isn't that you cannot *validate*, but that the reference cannot correctly *link*. The validation error is only a consequence of the linker error.

**Reply by zrwsmd** [reply]
so i should override computeLocalScopes method?
this file actually separated into different files,1-10 lines is at aa.st,  below is at main.st
![image](https://github.com/eclipse-langium/langium/assets/26995040/43cfba1a-ff07-4b7f-97ce-8f989284a3e1)

**Reply by msujew** [reply]
Yes, if the element should be visible to the elements of the same file, you shouls override the `computeLocalScopes` method.

**Reply by zrwsmd** [reply]
it works out!very thanks.if i want to reference  `productId`,i guess also should define a class extends `DefaultScopeProvider`?

**Reply by msujew** [reply]
> if i want to reference productId,i guess also should define a class extends DefaultScopeProvider?

Yes, exactly. The `ScopeComputation` is in charge of the lexical scope (i.e. which elements are visible on a block-based level), while the `ScopeProvider` allows you to define arbitrary linking rules for any reference.
